tiledb/common/evaluator/evaluator.h

```diff
@@ -0,0 +1,249 @@
+   * recently used element from the cache.
+   */
+  List lru_;
+
+  /**
+   * Each entry stores a promised value which is in the process of
+   * being fetched. Readers will wait on it to become available.
+   */
+  std::unordered_map<Key, OndemandValue> inprogress_;
+
+  /**
+   * Map of cached entries by key. Each entry stores and iterator
+   * within the LRU list and a shared_ptr to the value.
+   * If an entry is in this map, it is guaranteed that the value is
+   * already available to serve readers.
+   */
+  std::unordered_map<Key, Cached_Entry> entries_;
+
+  /** Bookkeeping lock */
+  std::mutex mutex_;
+};
+
+template <class Key, class Value>
+class NoCachePolicy {
+ public:
+  using key_type = Key;
+  using value_type = Value;
+
+  NoCachePolicy() = default;
+
+  template <class F>
+  std::shared_ptr<Value> operator()(F f, Key key) {
+    return std::get<1>(f(key));
+  }
+};
+
+template <class Key, class Value, size_t N = 0>
+class MaxEntriesCache : public CachePolicyBase<Key, Value> {
+ public:
+  MaxEntriesCache()
+      : num_entries_(0) {
+  }
+
+  template <class F>
+  std::shared_ptr<Value> operator()(F f, Key key) {
+    return std::get<1>(f(key));
+  }
+
+  virtual std::shared_ptr<Value> enforce_policy(
+      [[maybe_unused]] size_t mem_usage) override {
+    if (num_entries_ == max_entries_) {
+      auto evicted_value = this->evict();
+      --num_entries_;
+      return evicted_value;
+    }
+    ++num_entries_;
+    return nullptr;
+  }
+
+ private:
+  size_t num_entries_;
+  static const size_t max_entries_ = N;
+};
+
+template <class Key, class Value, size_t N = 1024>
+class MemoryBudgetedCache : public CachePolicyBase<Key, Value> {
+ public:
+  MemoryBudgetedCache()
+      : memory_consumed_(0) {
+  }
+
+  template <class F>
+  std::shared_ptr<Value> operator()(F f, Key key) {
+    return std::get<1>(f(key));
+  }
+
+  virtual std::shared_ptr<Value> enforce_policy(size_t mem_usage) override {
+    if (memory_consumed_ + mem_usage > memory_budget_) {
+      auto evicted_value = this->evict();
+      memory_consumed_ -= mem_usage;
+      return evicted_value;
+    }
+
+    memory_consumed_ += mem_usage;
+    return nullptr;
+  }
+
+ private:
+  /** The maximum amount of bytes managed by this cache */
+  static const size_t memory_budget_ = N;
+
+  /** The maximum amount of bytes currently consumed by this cache */
+  size_t memory_consumed_;
+};
+
+template <class Policy, class F>
+class Evaluator {
+ public:
+  using Key = typename Policy::key_type;
+  using Value = typename Policy::value_type;
+
+  Evaluator()
+      : caching_policy_{} {
+  }
+
+  /**
+   * Evaluate callback on key and return the value.
+   * According to the policy specified, the value might be fetched
+   * directly from the cache or the callback might be executed to fetch it.
+   */
+  std::shared_ptr<Value> operator()(F f, Key key) {
+    return caching_policy_.template operator()<F>(f, key);
+  }
+
+ private:
+  Policy caching_policy_;
```

Comment:
It's likely easier to derive the evaluator from its policy (or policies) with `protected` or `private` inheritance than it is to have a member variable.